samples/led-matrix-weather/Program.cs

```diff
@@ -0,0 +1,274 @@
+        static Stopwatch s_showLocalIp = null;
+        static string[] s_ips;
+        static bool s_networkAvailable = false;
+        static Weather s_client;
+        static OpenWeatherResponse s_weatherResponse;
+        static TimeZoneInfo s_timeZonePst = TimeZoneInfo.FindSystemTimeZoneById("America/Los_Angeles");
+
+        static void Main(string[] args)
+        {
+            if (args.Length != 1)
+            {
+                Console.WriteLine("Usage: led-matrix-weather <openweathermap.org API key>");
+                Console.WriteLine();
+                Console.WriteLine("If wrong key is used or there is no network connection");
+                Console.WriteLine("example data will be displayed.");
+                Console.WriteLine();
+                Console.WriteLine("For the first 30 seconds of execution");
+                Console.WriteLine("IP addresses will be displayed instead of data.");
+                return;
+            }
+
+            s_client = new Weather(args[0]);
+
+            UpdateWeather();
+            Task.Run(WeatherUpdater);
+            Task.Run(IpsGetter);
+
+            PinMapping mapping = PinMapping.MatrixBonnetMapping32;
+            RGBLedMatrix matrix = new RGBLedMatrix(mapping, 64, 64, 2, 2);
+
+            Task drawing = Task.Run(() =>
+            {
+                matrix.StartRendering();
+
+                while (true)
+                {
+                    Action<RGBLedMatrix> scenario = s_scenario;
+
+                    if (scenario == null)
+                        break;
+
+                    Stopwatch sw = Stopwatch.StartNew();
+                    scenario(matrix);
+
+                    if (s_scenario != null && sw.ElapsedMilliseconds < 100)
+                    {
+                        Debug.WriteLine("Scenario execution finished in less than 100ms. This is likely due to bug.");
+                    }
+                }
+            });
+
+            try
+            {
+                if (!Console.IsOutputRedirected)
+                {
+                    while (s_scenario != null)
+                    {
+                        switch (Console.ReadKey(intercept: true).Key)
+                        {
+                            case ConsoleKey.Q:
+                            {
+                                s_scenario = null;
+                                break;
+                            }
+                        }
+                    }
+                }
+
+                drawing.Wait();
+            }
+            finally
+            {
+                matrix.Dispose();
+            }
+        }
+
+        static void WeatherUpdater()
+        {
+            while (true)
+            {
+                UpdateWeather();
+                Thread.Sleep(5000);
+            }
+        }
+
+        static void IpsGetter()
+        {
+            try
+            {
+                s_ips = GetLocalNetworkIPAddresses().ToArray();

```

Comment:
Why do we want to run this in a separate thread? shouldn't it be possible to just run it on the main thread?

Reply:
That's what I initially did and it works most of the time but when running this when device is booting the network card might still be initializing which is causing an exception or hang causing the updating thread to freeze.